examples/Enumerators/Enumerators.dfy

```diff
@@ -0,0 +1,143 @@
+
+    var e: Enumerator<int> := new SeqEnumerator(numbers);
+    while (e.HasNext()) 
+      invariant e.Valid() && fresh(e.Repr)
+      decreases e.Decreases()
+    {
+      var element := e.Next();
+
+      print element, "\n";
+    }
+  }
+
+  method Example2() {
+    var first := [1, 2, 3, 4, 5];
+    var second := [6, 7, 8];
+    var e1 := new SeqEnumerator(first);
+    var e2 := new SeqEnumerator(second);
+    var e := new ConcatEnumerator(e1, e2);
+   
+    while (e.HasNext()) 
+      invariant e.Valid() && fresh(e.Repr)
+      decreases e.Decreases()
+    {
+      var element := e.Next();
+
+      print element, "\n";
+    }
+  }
+
+  method PrintWithCommas() {
+    var first := [1, 2, 3, 4, 5];
+    var e := new SeqEnumerator(first);
+   
+    while (e.HasNext()) 
+      invariant e.Valid() && fresh(e.Repr)
+      decreases e.Decreases()
+    {
+      var element := e.Next();
+
+      print element;
+      if e.HasNext() {
+        print ", ";
+      }
+    }
+    print "\n";
+  }
+
+  method MappingExample() {
+    var first := [1, 2, 3, 4, 5];
+    var e1 := new SeqEnumerator(first);
+    var e := new MappingEnumerator(x => x + 2, e1);
+   
+    var result: seq<int> := [];
+    while (e.HasNext()) 
+      invariant e.Valid() && fresh(e.Repr)
+      decreases e.Decreases()
+    {
+      var element := e.Next();
+
+      result := result + [element];
+    }
+    assert e.enumerated == Seq.Map(x => x + 2, first);
+  }
+
+  // Some examples showing the equivalence between enumerator
+  // operations and sequence operations.
+
+  method RoundTrip(s: seq<nat>) returns (result: seq<nat>)
+    ensures result == s
+  {
+    var e: Enumerator := new SeqEnumerator(s);
+    result := CollectToSeq(e);
+  }
+
+  method AddTwoToEach(s: seq<nat>) returns (result: seq<nat>)
+    ensures result == Seq.Map(x => x + 2, s)
+  {
+    var e := new SeqEnumerator(s);
+    var mapped := new MappingEnumerator(x => x + 2, e);
+    result := CollectToSeq(mapped);
+  }
+
+  method SetToSeq<T>(s: set<T>) returns (result: seq<T>)
+    ensures multiset(result) == multiset(s)
+  {
+    var e := new SetEnumerator(s);
+    result := CollectToSeq(e);
+  }
+
+  method Filter<T>(s: seq<T>, p: T -> bool) returns (result: seq<T>)
+    ensures result == Seq.Filter(p, s)
+  {
+    var e := new SeqEnumerator(s);
+    var filtered: FilteredEnumerator := new FilteredEnumerator(e, p);
+    result := CollectToSeq(filtered);
+    assert filtered.Valid();
+  }
+
+  method Concatenate<T>(first: seq<T>, second: seq<T>) returns (result: seq<T>)
+    ensures result == first + second
+  {
+    var e1 := new SeqEnumerator(first);
+    var e2 := new SeqEnumerator(second);
+    var concatenated: ConcatEnumerator := new ConcatEnumerator(e1, e2);
+    result := CollectToSeq(concatenated);
+    assert concatenated.Valid();
+  }
+
+  // TODO: The examples above work because Dafny is aware of the concrete
+  // types of the various enumerator values, and hence knows the additional post-conditions
```

Comment:
Where is this needed? For all `new` allocations in this `Demo` module, I gave the trait as an explicit type of the receiving variable and everything still verifies. For example, I changed the declarations of `e1`, `e2`, and `e` in `Example2()` to

``` dafny
    var e1: Enumerator := new SeqEnumerator(first);
    var e2: Enumerator := new SeqEnumerator(second);
    var e: Enumerator := new ConcatEnumerator(e1, e2);
```

Reply:
`Example2` is just fine with that because it doesn't assert anything about the behavior of the enumerators. But `Concatenate` has a post-condition that isn't verified if I remove the explicit type declaration on L119. That's because it needs the extra post-condition implied by `!ConcatEnumerator.HasNext()`, so Dafny doesn't have enough information if I don't guide it to recognize the concrete type of that enumerator. I'm trying to figure out if there's a better solution there.

Reply:
Ah don't spend any further thought on this, I have a solution that seems to work nicely. Will push something tomorrow.